Mathlib/Data/Matroid/Circuit.lean

```diff
@@ -319,4 +336,128 @@ lemma ext_iff_circuit {M₁ M₂ : Matroid α} :
     M₁ = M₂ ↔ M₁.E = M₂.E ∧ ∀ C, M₁.Circuit C ↔ M₂.Circuit C :=
   ⟨fun h ↦ by simp [h], fun h ↦ ext_circuit h.1 fun C hC ↦ h.2 (C := C)⟩
 
+section Elimination
+
+/-! ### Circuit Elimination -/
+
+variable {ι : Type*} (x : ι → α) {z : α} {C₁ C₂ : Set α}
+
+/-- A version of `Matroid.Circuit.strong_multi_elimination` that is phrased using insertion. -/
+lemma strong_multi_elimination_insert (I : ι → Set α) (hxI : ∀ i, x i ∉ I i) {J : Set α}
+    (hC : ∀ i, M.Circuit (insert (x i) (I i))) (hJx : M.Circuit (J ∪ range x)) (hzJ : z ∈ J)
+    (hzI : ∀ i, z ∉ I i) : ∃ C' ⊆ J ∪ ⋃ i, I i, M.Circuit C' ∧ z ∈ C' := by
+  -- we may assume that `ι` is nonempty, and it suffices to show that
+  -- `z` is spanned by the union of the `I` and `J \ {z}`.
+  obtain hι | hι := isEmpty_or_nonempty ι
+  · exact ⟨J, by simp, by simpa [range_eq_empty] using hJx, hzJ⟩
+  suffices hcl : z ∈ M.closure ((⋃ i, I i) ∪ (J \ {z})) by
+    rw [mem_closure_iff_exists_circuit_of_not_mem (by simp [hzI])] at hcl
+    obtain ⟨C', hC'ss, hC', hzC'⟩ := hcl
+    refine ⟨C', ?_, hC', hzC'⟩
+    rwa [union_comm, ← insert_union, insert_diff_singleton, insert_eq_of_mem hzJ] at hC'ss
+  replace hC := show ∀ (i : ι), M.closure (I i) = M.closure ({x i} ∪ (I i))
+    by simpa [diff_singleton_eq_self (hxI _)] using fun i ↦ (hC i).closure_diff_singleton_eq (x i)
+  -- This is true because each `I i` spans `x i` and `(range x) ∪ (J \ {z})` spans `z`.
+  rw [closure_union_congr_left <| closure_iUnion_congr _ _ hC, iUnion_union_distrib,
+    iUnion_singleton_eq_range, union_right_comm]
+  refine mem_of_mem_of_subset (hJx.mem_closure_diff_singleton_of_mem (.inl hzJ))
+    (M.closure_subset_closure (subset_trans ?_ subset_union_left))
+  rw [union_diff_distrib, union_comm]
+  exact union_subset_union_left _ diff_subset
+
+/-- A generalization of the strong circuit elimination axiom `Matroid.Circuit.strong_elimination`
+to an infinite collection of circuits.
```

Comment:
Is it possible to describe in words what the statement is saying here? `C` is the infinite collection of circuits, I guess, but it's not clear to me what the role of `C₀` is (informally speaking). Can we also do the same for `strong_multi_elimination_insert`? (For instance, are `I` and `C` playing the same role?)